src/psyclone/domain/common/transformations/invoke_trans.py

```diff
@@ -0,0 +1,238 @@
+                "argument contains a CodeBlock with content "
+                "({1}) which is not a StructureConstructor."
+                "".format(self.name, type(fp2_node).__name__))
+
+    def validate(self, call, options=None):
+        '''Validate the call argument.
+
+        :param call: a PSyIR call node capturing an invoke call in \
+            generic PSyIR.
+        :type call: :py:class:`psyclone.psyir.nodes.Call`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :raises TransformationError: if the supplied call argument is \
+            not a PSyIR Call node.
+        :raises TransformationError: if the supplied call argument \
+            does not have the expected name which would identify it as an \
+            invoke call.
+        :raises TransformationError: if the invoke arguments are not a \
+            PSyIR ArrayReference or CodeBlock.
+
+        '''
+        if not isinstance(call, Call):
+            raise TransformationError(
+                "Error in {0} transformation. The supplied call argument "
+                "should be a `Call` node but found '{1}'."
+                "".format(self.name, type(call).__name__))
+        if not call.routine.name.lower() == self._invoke_name:
+            raise TransformationError(
+                "Error in {0} transformation. The supplied call argument "
+                "should be a `Call` node with name '{1}' but found '{2}'."
+                "".format(self.name, self._invoke_name, call.routine.name))
+        for arg in call.children:
+            if isinstance(arg, ArrayReference):
+                pass
+            elif isinstance(arg, CodeBlock):
+                for fp2_node in arg._fp2_nodes:
+                    self._validate_fp2_node(fp2_node)
+            else:
+                raise TransformationError(
+                    "Error in {0} transformation. The arguments to this "
+                    "invoke call are expected to be a CodeBlock or an "
+                    "ArrayReference, but found '{1}'."
+                    "".format(self.name, type(arg).__name__))
+
+    def apply(self, call, options=None):
+        ''' Apply the transformation to the supplied node.
+
+        :param call: a PSyIR call node capturing an invoke call in \
+            generic PSyIR.
+        :type call: :py:class:`psyclone.psyir.nodes.Call`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        '''
+        self.validate(call, options=options)
+
+        kernel_calls = []
+        for call_arg in call.children:
+            arg_info = []
+            if isinstance(call_arg, ArrayReference):
+                # Structure constructor mis-parsed as an array
+                # reference.
+                type_symbol = call_arg.symbol
+                args = call_arg.children
```

Comment:
If you do call_arg.pop_all_children() here as well then you will not need the `args = [arg.copy() for arg in args]` below

Reply:
Great!